### PR TITLE
Double delta compression fixed merge

### PR DESCRIPTION
**Pull request by ashtul**
Replaces #236

### CODE REVIEW

**Comment by ashtul** [comment]
@danni-m @gkorland 
I have separated the append and read parts from the chunk itself for encapsulation.

File: src/tsdb.c [hunk 1]
```diff
@@ -45,20 +46,23 @@ void SeriesTrim(Series * series) {
...
     void *currentKey;
     size_t keyLen;
     timestamp_t minTimestamp = series->lastTimestamp > series->retentionTime ?
     		series->lastTimestamp - series->retentionTime : 0;
 
     while ((currentKey=RedisModule_DictNextC(iter, &keyLen, (void*)&currentChunk)))
     {
-        if (ChunkGetLastTimestamp(currentChunk) < minTimestamp)
+//        if (ChunkGetLastTimestamp(currentChunk) < minTimestamp)
+        if (CChunk_GetLastTimestamp(currentChunk) < minTimestamp)
```
Comment (danni-m) [inline src/tsdb.c]: @ashtul please use the original Chunk API + add a makefile switch between compressed chunk and non-compressed chunk.

File: src/tsdb.h [hunk 2]
```diff
@@ -24,7 +25,8 @@ typedef struct CompactionRule {
 
 typedef struct Series {
     RedisModuleDict* chunks;
-    Chunk* lastChunk;
+//    Chunk* lastChunk;
```
Comment (danni-m) [inline src/tsdb.h]: @ashtul you will have an isusue here because `Chunk*` should be some sort of a generic type, you can look at `redismodule.h` and see how redis declares a non-complete struct for example:
```
typedef struct RedisModuleDict RedisModuleDict;
```

File: tests/test_module.py [hunk 3]
```diff
@@ -159,9 +159,9 @@ def test_sanity(self):
 
             expected_result = {
               'totalSamples': 1500L,
-              'memoryUsage': 29022L,
+              'memoryUsage': 4238L,
```
Comment (danni-m) [inline tests/test_module.py]: you will need to support both chunk types, take a look at my branch to how I changed the testing to accommodate both.

File: src/module.c [hunk 4]
```diff
@@ -898,8 +900,7 @@ int TSDB_incrby(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
 
     double incrby = 0;
     if (RMUtil_ParseArgs(argv, argc, 2, "d", &incrby) != REDISMODULE_OK)
-        return RedisModule_WrongArity(ctx);
-
+        return RedisModule_ReplyWithError(ctx, "TSDB: invalid increase/decrease value");
```
Comment (danni-m) [inline src/module.c]: Good catch, should be extracted to a different PR.

File: src/gorilla.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#ifndef GORILLA_H
+#define GORILLA_H
+
+#include <sys/types.h>      // u_int_t
+#include <stdbool.h>        // bool
+
+#define timestamp_t u_int64_t
```
Comment (danni-m) [inline src/gorilla.h]: Did you move timestamp_t to here or did you redfined it?

File: src/compressed_chunk.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
+/*
+* Copyright 2018-2019 Redis Labs Ltd. and Contributors
+*
+* This file is available under the Redis Labs Source Available License Agreement
+*/
+
+#ifndef COMPRESSED_CHUNK_H
+#define COMPRESSED_CHUNK_H
+
+#include <sys/types.h>      // u_int_t
```
Comment (danni-m) [inline src/compressed_chunk.h]: small comment, no need for the verbose comments to why you are importing a specific header, those things change a lot and pretty standard.

File: src/gorilla.c [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    u_int64_t rv = 0;
+    u_int8_t idx64 = *idx % _64BIT;
+    u_int8_t available = _64BIT - idx64;
+    u_int64_t *runner = &arr[*idx / _64BIT];
+
+    if(available >= dataLen) {
+        rv = clearBits(*runner >> idx64, dataLen);
+    } else {
+        u_int8_t left = dataLen - available;
+        rv = clearBits(*runner >> idx64, available);
```
Comment (danni-m) [inline src/gorilla.c]: I might be wrong, but should the first x bits be shifted by `left` and the rest should be just be used with binary or?

File: src/gorilla.c [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+  chunk->prevTimestamp = timestamp;
+  return CC_OK;
+}
+
+static int appendV(CompressedChunk *chunk, double value) {
+  union64bits val;
+  val.d = value;
+  u_int64_t xorWithPrevious = val.u ^ chunk->prevValue.u;
+
+  // Checked for 1 bit in appendTS
```
Comment (danni-m) [inline src/gorilla.c]: why check the +1 bit in appendTS instead of here?

File: src/gorilla.c [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+static int64_t bin2int(u_int64_t x, u_int8_t l) {
+    bool negative = x & (1LL << (l - 1));
+    int64_t rv = x;
+    if (negative) {
+        rv -= 1LL << l;
+    }
+    return rv;
+}
+
+// Return `true` if byte is off
```
Comment (danni-m) [inline src/gorilla.c]:
```suggestion
// Return `true` if bit is off
```

File: src/gorilla.c [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+static bool isBitOff(u_int64_t *x, u_int64_t idx) {
```
Comment (danni-m) [inline src/gorilla.c]: Its worth calling this isFirstBitOff

File: src/gorilla.c [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+    iter->prevLeading = leading;
+    iter->prevTrailing = trailing;    
+  }
+
+  rv.u = xorValue ^ iter->prevValue.u;
+  return iter->prevValue.d = rv.d;
+}
+
+int CChunk_ReadNext(CChunk_Iterator *iter, u_int64_t *timestamp, double *value) {
+  *timestamp = iter->prevTS;
```
Comment (danni-m) [inline src/gorilla.c]: This can be confusing, there's basically two options here:
1. index == 0 and we should use base TS and base Val
2. index > 0 we should call readTS + readV

The current code suggest you return the previous sample call to read the next one but do not return it.

File: src/module.c (hunk 4 above)
Comment (ashtul) [reply src/module.c]: PR #260

File: src/tsdb.c (hunk 1 above)
Comment (ashtul) [reply src/tsdb.c]: @danni-m @gkorland 
I'm already working on the two chunks implementation.

Functions struct will have the function name without the chunk type (ex. AddSample, GetLastTimestamp...) while the functions themself will get a prefix of `Chunk_` or `CompressedC_`.
This will be done at the end after everything works.
Sounds good?

File: src/tsdb.h (hunk 2 above)
Comment (ashtul) [reply src/tsdb.h]: Will use Chunk_t which is `typedef void Chunk_t`

File: src/gorilla.h (hunk 5 above)
Comment (ashtul) [reply src/gorilla.h]: redefined since neither of gorilla.c or gorilla.h inherits from const.h

File: src/gorilla.c (hunk 7 above)
Comment (ashtul) [reply src/gorilla.c]: @danni-m 
I don't think I follow.
Can you write it as a code?

File: src/gorilla.c (hunk 8 above)
Comment (ashtul) [reply src/gorilla.c]: Since 1 is the minimum space required for the value.
If when writing the timestamp to the chunk there are no bits left, we will create a new one. If there is 1 bit, there is a chance (maybe even a good chance, depends on data), that it will be sufficient.
Otherwise, we will write both to the new chunk.

File: src/gorilla.c (hunk 11 above)
Comment (ashtul) [reply src/gorilla.c]: This is being handled at `SeriesIteratorGetNext`

File: src/gorilla.c (hunk 10 above)
Comment (ashtul) [reply src/gorilla.c]: It isn't just for the first bit but for any bit

File: src/gorilla.h [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  u_int64_t prevTS;
+  int64_t prevDelta;
+
+  // value vars
+  union64bits prevValue;  
+  u_int8_t prevLeading;
+  u_int8_t prevTrailing;
+} CChunk_Iterator;
+
+enum result {
```
Comment (rafie) [inline src/gorilla.h]: Uppercase Result. Also, since we don't have namespaces, it might get in conflict with other headers, so we may want to add some prefix.

Comment (rafie) [reply src/gorilla.h]: Move it to const.h and rename enum to Chunk_Result.

**Comment by rafie** [comment]
We can add the article link as a comment.

File: src/gorilla.h (hunk 12 above)
Comment (ashtul) [reply src/gorilla.h]: > Move it to const.h and rename enum to Chunk_Result.

Done in commit https://github.com/RedisTimeSeries/RedisTimeSeries/pull/261/commits/383d425766245639d42c4756dcd111faf38c6d61

**Comment by ashtul** [comment]
@rafie please write further comments on PR #261

**Comment by ashtul** [comment]
Current Double Delta PR is #261
